### PR TITLE
Remove ready directory which created in empty volumeMounter setUp func

### LINKED ISSUE #72675: configmap volume should delete the "ready" file of the internal emptydir volume on mount failure
**What happened**:
There is a node in my cluster whose disk is full on root partition, I cleared the disk and then rebooted the node to recover it. After the reboot, the kube-proxy on that node got stuck in "ContainerCreating" status and never changes, and all other pods can't run properly since the kubeproxy is not runing.

**What you expected to happen**:
Kubeproxy on that node can run normally after the disk clean and reboot

**How to reproduce it (as minimally and precisely as possible)**:
1. make the disk full on root partition
-  in my env, leave roughly 100k space on root partition would make this issue more reproducible

- this issue is kind of tricky, to reproduce it, one need to populate the disk repeatedly since some disk space would be cleared because of pod eviction

-  the kube-proxy got deleted and created repeatedly because of pod eviction and daemon-set mechanism, it would stuck in ContainerCreating status if the disk space is not enough for creating volume contents of kube-proxy configMap volume, the first pod error event is like:
_MountVolume.Setup failed for volume "kube-proxy":.....no space left on device_

- After it got stuck, the following  pod error event woud repeat over and over, kube-proxy is trying to mount configMap volume but always fail
_MountVolume.Setup failed for volume "kube-proxy": stat /var/lib/kubelet/pods/{pod_id}/volumes/kubernetes.io~configmap/kube-proxy no such file or directory_

2.clear the disk and reboot
- the reboot is neccesary to reproduce the issue, without the reboot the pod eviction would kick in to delete kube-proxy pod

3. now the kube-proxy woud stuck in ContainerCreating Status forever, restart kubelet has no effects on that, kube-proxy pod need to be mannully deleted to get recover
- kube-proxy pod can't mount  configMap volume, the following error keeps repeating
_MountVolume.Setup failed for volume "kube-proxy": stat /var/lib/kubelet/pods/{pod_id}/volumes/kubernetes.io~configmap/kube-proxy no such file or directory_

**Anything else we need to know?**:
After digging into this issue for a quite long time, i think the reason for this issue is:

The configMap volume uses emptyDir internally to create the volume directory, after the directory is created, configMap volume failed to populate it with configMap content because the disk is full, then configMap volume calls emptyDir.tearDown() to clear the temp directory.

The problem is that emptyDir uses a "ready" file under /var/lib/kubelet/pods/{podId}/plugin/..  to determine whether the directory is created or not, if the ready file is present, then it wouldn't create the directory. The tearDown method only remove the directory, it doen't remove the ready file, so after the first out of disk error, the diretory creation always get skipped, hence no such file or directory error repeat over and over.

**Environment**:
- Kubernetes version (use `kubectl version`):
v.1.11.5
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
cent os 7
- Kernel (e.g. `uname -a`):
- Install tools:
kubeadm
- Others:

### PR DESCRIPTION
**Pull request by changyaowei**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>

> /kind bug

**What this PR does / why we need it**:

For configMap volumeMounter , kubelet will  invoke emptyDir volumeMounter to do ```setUp``` and ```TearDown```. In emptyDir volumeMounter ```setUp``` func,  kubelet will check ```ready``` directory exist, if exist  and   If the medium is memory, and a mountpoint is present, then the volume will be thinked ready ,will not create volume directory again.
image

In configMap volumeMounter, it invoke emtpyDir volumeMount to  do ```setUp```  and then write payload and so on, if failed, it will invoke emtpyDir volumeMount to do ```TearDown```
image

In emptyDir volumeMounter ``` TearDown ```, it will check volume directory whether exist, if not exist ,it will do nothing, else it will remove volume directory, but not clean ready directory.
image

In scenario，in ```setUp``` func, after volume directory and  ready directory created, but write payload  error, it will clean  volume directory. In next loop, because of ready directory exist, it will not create volume directory again, and then ```NewAtomicWriter``` func  which is check volume directory first will error, just like 
```
Nov 06 16:18:18 9-13-2-186 kubelet[1275]: E1106 16:18:18.186965    1275 configmap.go:244] Error creating atomic writer: stat /var/lib/kubelet/pods/a8bfc7bd-daff-11e9-b117-fae200526c00/volumes/kubernetes.io~configmap/kube-proxy-master: no such file or directory
Nov 06 16:18:18 9-13-2-186 kubelet[1275]: W1106 16:18:18.187092    1275 empty_dir.go:373] Warning: Unmount skipped because path does not exist:
/var/lib/kubelet/pods/a8bfc7bd-daff-11e9-b117-fae200526c00/volumes/kubernetes.io~configmap/kube-proxy-master
Nov 06 16:18:18 9-13-2-186 kubelet[1275]: E1106 16:18:18.187294    1275 nestedpendingoperations.go:267] Operation for "\"kubernetes.io/configmap
/a8bfc7bd-daff-11e9-b117-fae200526c00-kube-proxy-master\" (\"a8bfc7bd-daff-11e9-b117-fae200526c00\")" failed. No retries permitted until 2019-11
```

So in emptyDir volumeMounter ``` TearDown ```, kubelet should clean ready directory.

**Which issue(s) this PR fixes**:

Fixes #72675 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Remove ready dir which created in setup func
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by jingxu97** [comment]
@changyaowei, thanks for the PR. Do you have time to add simple unit tests for this fix? Thanks!

**Comment by rata** [comment]
Sorry, I was away. I'll try to have a look during this week. Thanks for the patch! :-)

**Review by saad-ali — COMMENTED** [review]
I don't know this logic very well. But as I understand it, ready is a file not a directory. And it is cleaned up when the emptyDir is deleted. In this case it looks like you are adding logic to always delete the entire directory? Is the ready directory different from the emptyDir? Can you give an example.

Want to get @pmorie's thoughts since he is more familiar with this code.

**Comment by hickeng** [comment]
/remove-lifecycle rotten

Hit this behaviour just now on 1.18.2 - see [this comment](https://github.com/kubernetes/kubernetes/issues/72675#issuecomment-673694625).

**Comment by xing-yang** [comment]
@saad-ali has these comments:

> I don't know this logic very well. But as I understand it, ready is a file not a directory. And it is cleaned up when the emptyDir is deleted. In this case it looks like you are adding logic to always delete the entire directory? Is the ready directory different from the emptyDir? Can you give an example.

@pmorie and @jingxu97, can you please take a look?

File: pkg/volume/util/util.go [hunk 1]
```diff
@@ -108,6 +108,13 @@ func SetReady(dir string) {
 	file.Close()
 }
 
+// CleanReady remove the 'ready' file
+func CleanReady(dir string) {
+	if err := os.RemoveAll(dir); err != nil && !os.IsNotExist(err) {
```
Comment (xing-yang) [inline pkg/volume/util/util.go:113]: Instead of removing the entire dir, should we be removing only the `ready` file?

**Comment by jingxu97** [comment]
I dig some old code and found this PR added the logic about using readyFile https://github.com/kubernetes/kubernetes/pull/9384/files

My understanding is emptyDir setup requires creating a volume dir and also setting up the correct permission. Since it takes time, if another EmptyDir.setup is called, we cannot verify whether the setup is finished or not by just checking the volume dir. Instead it creates a separate readyDir after volume setup is done and use it to verify whether the volume is ready.  
So I think we should clean up readyDir during teardown. Otherwise, it will be left on the host forever.

File: pkg/volume/util/util.go (hunk 1 above)
Comment (jingxu97) [reply pkg/volume/util/util.go:113]: readyDir should be empty, I think. But to be safe, maybe we call os.Remove() and throw error if it is not empty.

**Comment by jingxu97** [comment]
cc @gnufied @saad-ali

**Comment by changshuchao** [comment]
@changyaowei Hello, I also encountered this problem. How is this problem progressing?

File: pkg/volume/util/util.go (hunk 1 above)
Comment (jingxu97) [reply pkg/volume/util/util.go:113]: actually getMetaDir contains one file named "ready", there should no other files. I think RemoveAll is fine.

**Pull request closed without merging**